neymarjimoh/react-movie-app
# created banner, row and custom button components for the home page

Dynamic application home page which consists of a row, banner and custom button components.

To enable the on demand movie application have a dynamic home page, i made multiple requests to 'the movie database' (tmdb) API to fetch various title categories such as 'Trending Now', 'Netflix Originals', 'Documentaries' and others which serve are rows in the home page. This request was made using axios, therefore i added axios using 'yarn add axios' to my cloned project.

To start off development, i created a single row functional component which would serve for multiple title categories, a  single h2 tag and div are the jsx elements which make up the row function. A title prop is passed to the row component from the home page component which is then written in curly braces inside the h2 tag which displays the title of each row.

Secondly, using react hooks i set up a movies state and setMovies with the useState being an empty array, then i called useEffect and created an async function called fetchData which uses axios to fetch the above mentioned film title categories in a request variable. setMovies is then called with the request variable in it to fetch its data.results params. This sets the state of 'movies with an array of different title categories and other params.

Movies is then placed in the div tag and mapped through and params such as the poster_path and backdrop_path is used to create an image for each movie title displayed in each row/category.

To display the movie trailers when a movie is clicked, i used two packages called 'react-youtube' and 'movie-trailer' to achieve this. You can add these packages using 'yarn add react-youtube' and 'yarn add movie-trailer' to your project.

Asides the h2 and div tags. i embedded js after these tags to display the movie trailer using the react-youtube tag and set a trailer and setTrailer to useState with an  empty string. If trailer became true, the movie trailer displayed which was handled by the a handleClick event function on the movie posters in the above div. 

Next i created the banner, which consists of a background image in the header tag, a div containing the movie title in a h1 tag, the banner buttons and a summary  tag.

To ensure banner images changed at random when the component mounts, i used useState hook in the same way as in the row component however when calling setMovie, i leveraged Math.floor and Math.random to help change the results from the data.results array fetched which params were also passed into the h1 tag to display the movie title and movie summary.

For the movie overview, a truncate function is used to summarize the overview if above 150 characters and an ellipse is used after such summaries.

I then imported some custom buttons for the 'Play' and 'My list' buttons.

How to test this feature: 

- Clone the branch to your machine
- Add axios, react-youtube, movie-trailer packages to your project
- Add yarn or npm to your project (Google it)
- Type yarn start in your code editor terminal or command line to launch the app
- Reload the page to see if the banner image changes
- Click on movie titles to watch trailers

N/B: Not all movies have trailers as the API do not provide that for developer/free version.

If you prefer to use sass for your css, sometimes the above packages added may have unmet dependencies with node-sass so you might have to remove your yarn.lock and node_modules and re-install them.
